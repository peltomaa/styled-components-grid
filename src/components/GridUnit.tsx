import * as React from 'react';
import styled from 'styled-components';
// @ts-ignore 
import createComponentFromTagProp from 'react-create-component-from-tag-prop';
import { BreakpointValues, Size, Visible } from '../types';
import gridUnit from '../mixins/gridUnit';

export type GridUnitProps = {
  size?: BreakpointValues<Size>,
  visible?: BreakpointValues<Visible>,
};

const GridUnitComponent: React.ComponentType<GridUnitProps> = createComponentFromTagProp({
  tag: 'div',
  prop: 'component',
  propsToOmit: ['width', 'visible']
});

const GridUnit = styled(GridUnitComponent) `
  box-sizing: border-box;
  ${gridUnit}
` as typeof GridUnitComponent;


export default GridUnit;
