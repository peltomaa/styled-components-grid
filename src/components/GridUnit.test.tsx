// @flow
import 'jest-styled-components';
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import GridUnit from './GridUnit';

describe('GridUnit', () => {

  it(`should render size`, () => {
    const element = renderer.create(<GridUnit size={0.3333} />).toJSON();
    expect(element).toHaveStyleRule('flex-basis', '33.33%');
    expect(element).toMatchSnapshot();
  });

  it(`should render size as "min"`, () => {
    const element = renderer.create(<GridUnit size="min" />).toJSON();
    expect(element).toHaveStyleRule('flex-grow', '0');
    expect(element).toMatchSnapshot();
  });

  it(`should render size as "max"`, () => {
    const element = renderer.create(<GridUnit size="max" />).toJSON();
    expect(element).toHaveStyleRule('flex-grow', '1');
    expect(element).toMatchSnapshot();
  });

  it('should render size at different breakpoints', () => {
    const element = renderer.create(<GridUnit size={{ mobile: 1, tablet: 0.5, desktop: 0 }} />).toJSON();
    expect(element).toHaveStyleRule('flex-basis', '100%');
    expect(element).toHaveStyleRule('flex-basis', '50%', { media: '(min-width:46.0625em)' });
    expect(element).toHaveStyleRule('flex-basis', '0%', { media: '(min-width:64.0625em)' });
    expect(element).toMatchSnapshot();
  });

  it(`should render visible as "true"`, () => {
    const element = renderer.create(<GridUnit visible={true} />).toJSON();
    expect(element).toHaveStyleRule('display', 'flex');
  });

  it(`should render visible as "false"`, () => {
    const element = renderer.create(<GridUnit visible={false} />).toJSON();
    expect(element).toHaveStyleRule('display', 'none');
  });

  it('should render visible at different breakpoints', () => {
    const element = renderer.create(<GridUnit visible={{ mobile: true, tablet: false, desktop: true }} />).toJSON();
    expect(element).toHaveStyleRule('display', 'flex');
    expect(element).toHaveStyleRule('display', 'none', { media: '(min-width:46.0625em)' });
    expect(element).toHaveStyleRule('display', 'flex', { media: '(min-width:64.0625em)' });
  });

});
