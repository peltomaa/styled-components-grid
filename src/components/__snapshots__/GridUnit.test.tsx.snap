// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridUnit should render size 1`] = `
.c0 {
  box-sizing: border-box;
  box-sizing: border-box;
  -webkit-flex-basis: 33.33%;
  -ms-flex-preferred-size: 33.33%;
  flex-basis: 33.33%;
  max-width: 33.33%;
}

<div
  className="c0"
  size={0.3333}
/>
`;

exports[`GridUnit should render size as "max" 1`] = `
.c0 {
  box-sizing: border-box;
  box-sizing: border-box;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  width: auto;
  max-width: none;
  max-width: 100%;
}

<div
  className="c0"
  size="max"
/>
`;

exports[`GridUnit should render size as "min" 1`] = `
.c0 {
  box-sizing: border-box;
  box-sizing: border-box;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  width: auto;
  max-width: none;
}

<div
  className="c0"
  size="min"
/>
`;

exports[`GridUnit should render size at different breakpoints 1`] = `
.c0 {
  box-sizing: border-box;
  box-sizing: border-box;
  -webkit-flex-basis: 100%;
  -ms-flex-preferred-size: 100%;
  flex-basis: 100%;
  max-width: 100%;
  -webkit-flex-basis: 100%;
  -ms-flex-preferred-size: 100%;
  flex-basis: 100%;
  max-width: 100%;
}

@media (min-width:46.0625em) {
  .c0 {
    -webkit-flex-basis: 50%;
    -ms-flex-preferred-size: 50%;
    flex-basis: 50%;
    max-width: 50%;
  }
}

@media (min-width:64.0625em) {
  .c0 {
    -webkit-flex-basis: 0%;
    -ms-flex-preferred-size: 0%;
    flex-basis: 0%;
    max-width: 0%;
  }
}

<div
  className="c0"
  size={
    Object {
      "desktop": 0,
      "mobile": 1,
      "tablet": 0.5,
    }
  }
/>
`;
